core/src/bytecode/ast/builder.rs

```diff
@@ -0,0 +1,716 @@
+//! A builder interface for producing Nickel values.
+//!
+//! Using this module, you can define Nickel values using a builder style. For example:
+//!
+//! ```rust
+//! # use nickel_lang_core::bytecode::ast::{record::MergePriority, Ast, Node, builder::Record, AstAlloc};
+//!
+//! let alloc = AstAlloc::new();
+//!
+//! let record = Record::new()
+//!     .field("foo")
+//!     .priority(MergePriority::Bottom)
+//!     .doc("foo?")
+//!     .not_exported()
+//!     .value(alloc.string("foo"));
+//!
+//! let record = record
+//!     .field("bar")
+//!     .value(Node::Number(42.into()))
+//!     .build(&alloc);
+//!
+//! // Gives {foo | doc "foo?" | not_exported | defaut = "foo", bar = 42}
+//! ```
+//!
+//! This modules also offers a simpler interface for basic values where all arguments can be
+//! provided at once, to avoid useless intermediate allocations.
+use indexmap::IndexMap;
+
+use crate::identifier::{Ident, LocIdent};
+
+use super::{
+    combine::Combine,
+    record::{FieldMetadata, MergePriority},
+    typ::Type,
+    *,
+};
+
+use std::rc::Rc;
+
+type StaticPath = Vec<Ident>;
+
+/// Typestate style tag for `Field`s that are not yet completely specified
+pub struct Incomplete();
+
+/// Typestate style tag for `Field`s that have been finalized
+pub struct Complete<'ast>(Option<Ast<'ast>>);
+
+/// A Nickel record field being constructed
+#[derive(Debug)]
+pub struct Field<'ast, State> {
+    /// This field stores different piece of information depending on the state of this field:
+    ///
+    /// - It's empty for an incomplete field (`()`).
+    /// - When a value (or `None`) has been set, `State` becomes [Complete] and stores the
+    ///   optional value of this field.
+    /// - Finally, when the field has been attached to a record, `state` store the parent record.
```

Comment:
The "Finally" here confused me for a bit, because it made me thing that the state lifecycle goes incomplete -> complete -> parent record but this is not the case.

Reply:
I think my understanding is the same as what you wrote, which is that the state can go Incomplete -> Complete (and then gets consumed by the record, but it doesn't become a `Field<Record>`) or Record -> Complete. But the wording of the comment initially made me think that the lifecycle went `Field<Incomplete> -> Field<Complete> -> Field<Record>`. Maybe it isn't important, it just tripped me up a little while reading...

Reply:
Fair enough, I'll try to reword the explanation to make it clearer :+1: